django/template/defaulttags.py

```diff
@@ -1167,6 +1167,54 @@ def now(parser, token):
     return NowNode(format_string, asvar)
 
 
+@register.simple_tag(takes_context=True)
+def query_string(context, query_dict=None, **kwargs):
+    """
+    Add, remove, and change parameters of the request's `QueryDict`,
+    and return the result as a querystring.
+
+    For example::
+
+        {% query_string foo=3 %}
+
+    To remove a key::
+
+        {% query_string foo=None %}
+
+    To use with pagination::
+
+        {% query_string page=page_obj.next_page_number %}
+
+    A custom `QueryDict` can also be used:
+
+        {% query_string my_query_dict foo=3 %}
+    """
+    if query_dict:
+        query_dict = query_dict.copy()
+    else:
+        query_dict = context.request.GET.copy()
+    keys_to_remove = []
+    for key, value in kwargs.items():
+        if value is None:
+            keys_to_remove.append(key)
+        try:
+            iter(value)
+        except TypeError:
+            query_dict[key] = value
+        else:
+            if isinstance(value, str):
+                query_dict[key] = value
+            else:
+                query_dict.setlist(key, value)
+    for key in keys_to_remove:
+        if key in query_dict:
+            del query_dict[key]
+    query_string = query_dict.urlencode()
+    if not query_string:
+        return ""
+    return f"?{query_string}"
```

Comment:
Don't want to bike shed... Can I just ask @knyghty+@felixxm to confirm we **do want** the `?` automatically prefixed? 

Elsewhere we handle `query_dict.urlencode()` without the prefix. 

(Not a biggie, not expressing a negative, just asking for a double check 🎁)

Reply:
I didn't think about it _too_ hard but I do find it nice that a user can just do this:

```html+django
<a href="{% url "foo" %}{% query_string %}">something</a>
```

without needing to know if the query string is empty or not.

Reply:
Yep, that makes sense. 

(`/foo/?` would be fine though too right? But maybe not as nice.)

Reply:
It would but indeed I think people (including me) would be a bit annoyed at this. It also adds a potential footgun where you don't realise you need to do this and you end up with `/foo/color=blue&size=M`. The other way around, if you double your `??` seems to be handled (by the browser?) as when I make requests like this the duplicate `?` is removed.